src/event_store/in_memory_event_store.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
+//
+// This program is licensed to you under the Apache License Version 2.0,
+// and you may not use this file except in compliance with the Apache License Version 2.0.
+// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the Apache License Version 2.0 is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+
+use crate::event_batch::EventBatch;
+use crate::event_store::EventStore;
+use crate::payload::{Payload, PayloadBuilder};
+use crate::Error;
+
+// This is pub(crate) as it is used in BatchEmitter
+pub(crate) const DEFAULT_EVENT_STORE_CAPACITY: usize = 10_000;
+const DEFAULT_BATCH_SIZE: usize = 50;
+
+struct InMemoryEventStoreQueue {
+    queue: Vec<PayloadBuilder>,
+    capacity: usize,
+}
+
+// A slightly extended Vec to store maximum capacity,
+// along with returning an error on add if the maximum capacity is reached.
+impl InMemoryEventStoreQueue {
+    fn new(capacity: usize) -> Self {
+        InMemoryEventStoreQueue {
+            // `with_capacity` allocates `capacity` elements, to avoid later reallocation
+            queue: Vec::with_capacity(capacity),
+            capacity,
+        }
+    }
+
+    /// Add a payload to the queue
+    /// Returns an error if the queue is full
+    fn push(&mut self, payload: PayloadBuilder) -> Result<(), Error> {
+        if self.queue.len() == self.queue.capacity() {
+            return Err(Error::EventStoreError("Event store is full".to_string()));
+        }
+        self.queue.push(payload);
+        Ok(())
+    }
+}
+
+/// An implementation of the [EventStore] trait, that queues events in a Vec
+pub struct InMemoryEventStore {
+    event_queue: InMemoryEventStoreQueue,
+    batch_size: usize,
+}
+
+/// Provides an instance of [InMemoryEventStore], with the default batch size of 50, and a queue capacity of 10,000
+impl Default for InMemoryEventStore {
+    fn default() -> Self {
+        Self {
+            event_queue: InMemoryEventStoreQueue::new(DEFAULT_EVENT_STORE_CAPACITY),
+            batch_size: DEFAULT_BATCH_SIZE,
+        }
+    }
+}
+
+impl InMemoryEventStore {
+    pub fn new(queue_capacity: usize, batch_size: usize) -> Self {
+        Self {
+            event_queue: InMemoryEventStoreQueue::new(queue_capacity),
+            batch_size,
+        }
+    }
+
+    fn event_batch(&mut self, size: usize) -> Result<EventBatch, Error> {
+        if self.event_queue.queue.is_empty() {
+            return Err(Error::EventStoreError("Event store is empty".to_string()));
+        }
+
+        if size > self.batch_size {
+            return Err(Error::EventStoreError(
```

Comment:
I think that this check might be problematic when `emitter.flush()` is called. My understanding is that in that case, the emitter requests a `batch_of` with the full size of the event queue. This is likely to be larger than the batch size.

The question is what should happen in that case (when there are more events in queue than `batch_size`)? I can see two options:

1. just make one huge request with all the events currently in the event queue (I think that the Java tracker does that)
2. make N requests each with the size of up to `batch_size`

I think option 2 is a bit nicer than making a single request with an unbounded number of events, but probably option 1 would be fine for most cases.

Reply:
Now I realized that the number of events in the event store can never be larger than the batch size because we currently do not support retries. Once we introduce retries, I think this concern will become valid because it might be possible for the event store to become larger than `batch_size` in case batches fail to be sent out. But that is a concern for future us. Leaving this comment here just FYI, but feel free to ignore.